### PR TITLE
Performance Improvement - Avoid event copying for the first sub-pipeline

### PR DESCRIPTION
**Pull request by kkondaka**
### Description
Currently Pipeline connector code makes a copy of the records even if there is only one pipeline. Copying is required only when there are multiple sub pipelines.
For example if the config is like this -
`pipeline1 -> source -> processor -> pipeline-2(as sink) -> pipeline2 (source) -> processor -> sink`
This will do a copy of every record even though it is not necessary. 
This fix avoids copying to the first sink and makes a copy only when the same record is sent to multiple sinks.
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [X ] New functionality includes testing.
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [X ] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-api/src/main/java/org/opensearch/dataprepper/model/sink/Sink.java [hunk 1]
```diff
@@ -20,7 +20,7 @@
      *
      * @param records the records to write to the sink.
      */
-    void output(Collection<T> records);
+    void output(Collection<T> records, boolean copyFlag);
```
Comment (dlvenable) [inline data-prepper-api/src/main/java/org/opensearch/dataprepper/model/sink/Sink.java:23]: I don't think this is quite the right design. It is putting in a value for the `PipelineConnector` which other sinks will not need.

The `PipelineConnector` is a unique Sink and is internal to Data Prepper core.

Is there another design which can accomplish this goal without affecting the `Sink` interface or any other sinks?

**Review by dlvenable — CHANGES_REQUESTED** [review]
Thanks for working through these performance changes. I have some concerns about the design. Please see if you can solve this only changing data-prepper-core.

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/Pipeline.java [hunk 2]
```diff
@@ -255,9 +255,16 @@ private void shutdownExecutorService(final ExecutorService executorService, fina
...
         final List<Future<Void>> sinkFutures = new ArrayList<>(sinksSize);
-        router.route(records, sinks, (sink, events) ->
-                sinkFutures.add(sinkExecutorService.submit(() -> sink.output(events), null))
+        final List<DataFlowComponent<Sink>> firstSink = sinks.subList(0,1);
+        final List<DataFlowComponent<Sink>> remainingSinks = sinks.subList(1, sinksSize);
+        router.route(records, firstSink, (sink, events) ->
+                sinkFutures.add(sinkExecutorService.submit(() -> sink.output(events, false), null))
         );
+        if (remainingSinks.size() > 0) {
+            router.route(records, remainingSinks, (sink, events) ->
```
Comment (dlvenable) [inline data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/Pipeline.java:264]: This is calling the router twice which will require re-evaluating routes when using conditional routing. Thus, conditional routing sinks may have a performance hit.

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/Pipeline.java [hunk 3]
```diff
@@ -255,9 +255,16 @@ private void shutdownExecutorService(final ExecutorService executorService, fina
     List<Future<Void>> publishToSinks(final Collection<Record> records) {
         final int sinksSize = sinks.size();
         final List<Future<Void>> sinkFutures = new ArrayList<>(sinksSize);
-        router.route(records, sinks, (sink, events) ->
-                sinkFutures.add(sinkExecutorService.submit(() -> sink.output(events), null))
+        final List<DataFlowComponent<Sink>> firstSink = sinks.subList(0,1);
```
Comment (dlvenable) [inline data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/Pipeline.java:258]: This picks the first sink to have a copy. What if the first sink does not need the copy?

```
sink:
  - opensearch:
  - pipeline:
  - pipeline:
```

In that example, it will perform the copies anyway. So there is no performance gain.

Also, by using the first sink, you may also perform the copy if conditionally routing to the second sink. Consider the example below.

```
  sink:
    - pipeline:
        routes: [route-a]
    - pipeline:
        routes: [route-b]
```

This will perform the copy for all events which go to route-b even though they don't need to be copied for the performance gain.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/router/Router.java [hunk 4]
```diff
@@ -37,8 +42,23 @@ public <C> void route(
...
         final Map<Record, Set<String>> recordsToRoutes = routeEventEvaluator.evaluateEventRoutes(allRecords);
 
+        Set<Record> routedRecords = null;
+        /*
+         * If there are more than one sink and one of the sinks is
+         * pipeline connector, then we should make a copy of every
+         * record that is routed to more than one sink, so, to keep
+         * track of already routed records, initialize the set.
+         */
+        if (dataFlowComponents.size() > 1) {
```
Comment (dlvenable) [inline data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/router/Router.java:52]: Can we extract this out as well?

**Review by dlvenable — CHANGES_REQUESTED** [review]
Thanks! This is very close. With a little more work we should be able to have separation of concerns with this solution.

File: data-prepper-core/src/test/java/org/opensearch/dataprepper/pipeline/router/RouterCopyRecordStrategyTests.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+                                       .build();
+        recordsIn = IntStream.range(0, 10)
+                .mapToObj(i -> new Record<Event>(event))
+                .collect(Collectors.toList());
+        final RouterGetRecordStrategy getRecordStrategy = createObjectUnderTest(dataFlowComponents);
+        Record firstRecord = recordsIn.iterator().next();
+        Record recordOut = getRecordStrategy.getRecord(firstRecord);
+        assertThat(recordOut, equalTo(firstRecord));
+        recordOut = getRecordStrategy.getRecord(firstRecord);
+        assertThat(recordOut, not(equalTo(firstRecord)));
```
Comment (dlvenable) [inline data-prepper-core/src/test/java/org/opensearch/dataprepper/pipeline/router/RouterCopyRecordStrategyTests.java:117]: This would probably be better as `not(sameInstance(firstRecord))`.

In the future if we add support for `equals`/`hashCode` this would break and could be confusing.

File: data-prepper-core/src/test/java/org/opensearch/dataprepper/pipeline/router/RouterCopyRecordStrategyTests.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+                                       .withData(eventMap)
+                                       .withEventType("event")
+                                       .build();
+        recordsIn = IntStream.range(0, 10)
+                .mapToObj(i -> new Record<Event>(event))
+                .collect(Collectors.toList());
+        final RouterGetRecordStrategy getRecordStrategy = createObjectUnderTest(dataFlowComponents);
+        Record firstRecord = recordsIn.iterator().next();
+        Record recordOut = getRecordStrategy.getRecord(firstRecord);
+        assertThat(recordOut, equalTo(firstRecord));
```
Comment (dlvenable) [inline data-prepper-core/src/test/java/org/opensearch/dataprepper/pipeline/router/RouterCopyRecordStrategyTests.java:115]: I think you should use `sameInstance` to be very explicit on the behavior.

```
assertThat(recordOut, sameInstance(firstRecord));
```

This only works now because `equals`/`hashCode` is not yet implemented.

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/router/DataFlowComponentRouter.java [hunk 7]
```diff
@@ -5,81 +5,41 @@
...
+            recordsForComponent = Objects.isNull(getRecordStrategy) ? allRecords : getRecordStrategy.getAllRecords(allRecords);
         } else {
             recordsForComponent = new ArrayList<>();
             for (Record record : allRecords) {
                 final Set<String> routesForEvent = recordsToRoutes
                         .getOrDefault(record, Collections.emptySet());
 
                 if (routesForEvent.stream().anyMatch(dataFlowComponentRoutes::contains)) {
-                    recordsForComponent.add(getRecordForComponent(usedRecords, record));
+                    recordsForComponent.add(Objects.isNull(getRecordStrategy) ? record : getRecordStrategy.getRecord(record));
```
Comment (dlvenable) [inline data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/router/DataFlowComponentRouter.java:42]: I think we should require the strategy. It will simplify the code and allow you to have a consistent set of tests.

File: data-prepper-core/src/test/java/org/opensearch/dataprepper/pipeline/router/RouterTest.java [hunk 8]
```diff
@@ -190,9 +196,45 @@ void route_with_one_DataFlowComponents_And_PipelineConnector() {
             final DataFlowComponent<PipelineConnector> dataFlowComponent = mock(DataFlowComponent.class);
             pipelineDataFlowComponents.add(dataFlowComponent);
 
-            createObjectUnderTest().route(recordsIn, pipelineDataFlowComponents, pipelineComponentRecordsConsumer);
+            createObjectUnderTest().route(recordsIn, pipelineDataFlowComponents, null, pipelineComponentRecordsConsumer);
 
             verify(dataFlowComponentRouter).route(recordsIn, dataFlowComponent, recordsToRoutes, null, pipelineComponentRecordsConsumer);
         }
+
+        @Test
```
Comment (dlvenable) [inline data-prepper-core/src/test/java/org/opensearch/dataprepper/pipeline/router/RouterTest.java:204]: See my comments above about requiring the strategy.

You have added two tests which verify the behavior, but the unit tests themselves have many different scenarios.

The simplest approach is to mock the strategy at the top of the whole class and have it have the current implementation.

Put something like the following in the test setup.

```
getRecordStrategy = mock(RouterGetRecordStrategy.class);
when(getRecordStrategy.getRecord(any())).thenAnswer(a -> return a.getArguments(0));
when(getRecordStrategy.getAllRecords(anyCollection())).thenAnswer(a -> return a.getArguments(0));
```

Then you can remove these tests and use that one.

You can also add a `verify` behavior to ensure the strategy is called.

File: data-prepper-core/src/test/java/org/opensearch/dataprepper/pipeline/router/RouterTest.java [hunk 9]
```diff
@@ -190,9 +196,45 @@ void route_with_one_DataFlowComponents_And_PipelineConnector() {
...
+        void route_with_multiple_DataFlowComponents_And_Strategy() {
+            final DataFlowComponent dataFlowComponent = mock(DataFlowComponent.class);
+            dataFlowComponents = new ArrayList<>();
+            for (int i = 0; i < 5; i++) {
+                dataFlowComponents.add(dataFlowComponent);
+            }
+            RouterCopyRecordStrategy routerCopyRecordStrategy = new RouterCopyRecordStrategy(dataFlowComponents);
```
Comment (dlvenable) [inline data-prepper-core/src/test/java/org/opensearch/dataprepper/pipeline/router/RouterTest.java:211]: We want to remove these tests, but this is not a unit test - you are testing two classes now.

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/router/DataFlowComponentRouter.java [hunk 10]
```diff
@@ -5,81 +5,41 @@
...
-                if (usedRecords != null) {
-                    usedRecords.addAll(allRecords);
-                }
-            } else {
-                recordsForComponent = new ArrayList<>();
-                for (Record record : allRecords) {
-                    recordsForComponent.add(getRecordForComponent(usedRecords, record));
-                }
-            }
+            recordsForComponent = Objects.isNull(getRecordStrategy) ? allRecords : getRecordStrategy.getAllRecords(allRecords);
```
Comment (dlvenable) [inline data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/router/DataFlowComponentRouter.java:34]: Also, require the strategy.

File: data-prepper-core/src/test/java/org/opensearch/dataprepper/pipeline/router/RouterCopyRecordStrategyTests.java [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+    void setUp() {
+        recordsIn = IntStream.range(0, 10)
+                .mapToObj(i -> mock(Record.class))
+                .collect(Collectors.toList());
+    }
+
+    private <C> RouterGetRecordStrategy createObjectUnderTest(Collection<DataFlowComponent<C>> dataFlowComponents) {
+        return new RouterCopyRecordStrategy(dataFlowComponents);
+    }
+
```
Comment (dlvenable) [inline data-prepper-core/src/test/java/org/opensearch/dataprepper/pipeline/router/RouterCopyRecordStrategyTests.java:56]: Please add a test that verifies that the strategy passed into the constructor is not null.

```
strategy = null;
assertThrows(NullPointerException.class, this::createObjectUnderTest);
```

File: data-prepper-core/src/test/java/org/opensearch/dataprepper/pipeline/router/RouterTest.java (hunk 9 above)
Comment (dlvenable) [reply data-prepper-core/src/test/java/org/opensearch/dataprepper/pipeline/router/RouterTest.java:211]: If you want to cover that the copy works as expected, then this is an integration test. Please use the somewhat new integration test framework.

Here is a sample integration test:
https://github.com/opensearch-project/data-prepper/blob/main/data-prepper-core/src/integrationTest/java/org/opensearch/dataprepper/integration/Router_SingleRouteIT.java

I'd say this is optional.

**Review by dlvenable — CHANGES_REQUESTED** [review]
Thanks! It looks really good.

I think there a few other minor changes that will help the code here.

File: data-prepper-core/src/test/java/org/opensearch/dataprepper/pipeline/router/DataFlowComponentRouterTest.java [hunk 12]
```diff
@@ -51,14 +54,30 @@ private static class TestComponent {
...
     }
 
     private DataFlowComponentRouter createObjectUnderTest() {
         return new DataFlowComponentRouter();
     }
 
+    @Test
+    void route_with_null_strategy_throws() {
+        getRecordStrategy = null;
+        assertThrows(NullPointerException.class, () -> createObjectUnderTest().route(recordsIn, dataFlowComponent, null, getRecordStrategy, componentRecordsConsumer));
```
Comment (dlvenable) [inline data-prepper-core/src/test/java/org/opensearch/dataprepper/pipeline/router/DataFlowComponentRouterTest.java:70]: We should fail quickly. So you should check for a null pointer in the constructor and change this test to:

```
assertThrows(NullPointerException.class, () -> createObjectUnderTest());
```

In the current code and test, we will not know if there is an issue until after the pipeline is running without that null check.

**Review by dlvenable — CHANGES_REQUESTED** [review]

**Review by dlvenable — APPROVED** [review]

**Review by chenqi0805 — APPROVED** [review]
